openhands/memory/condenser/impl/browser_output_condenser.py

```
@@ -0,0 +1,41 @@
+from __future__ import annotations
+
+from openhands.core.config.condenser_config import BrowserOutputCondenserConfig
+from openhands.events.event import Event
+from openhands.events.observation import BrowserOutputObservation
+from openhands.events.observation.agent import AgentCondensationObservation
+from openhands.memory.condenser.condenser import Condenser
+
+
+class BrowserOutputCondenser(Condenser):
+    """A condenser that masks the observations from browser outputs outside of a recent attention window."""
+
+    def __init__(self, attention_window: int = 1):
+        self.attention_window = attention_window
+        super().__init__()
+
+    def condense(self, events: list[Event]) -> list[Event]:
+        """Replace the content of browser observations outside of the attention window with a placeholder."""
+        results: list[Event] = []
+        cnt: int = 0
+        for event in reversed(events):
+            if (
+                isinstance(event, BrowserOutputObservation)
+                and cnt >= self.attention_window
+            ):
+                results.append(AgentCondensationObservation('<MASKED>'))
```

Instead of using '<MASKED>', can we use something like `[Current URL: <event.url>]\n Content omitted` ? My intuition is URL is often useful and since it’s short anyways, we could still include it in the condensed observation

---

I second this -- without this change, I believe this is functionally equivalent to the `ObservationMaskingCondenser`.

Maybe we should combine the two and just have a match-case that picks how to do the masking based on the kind of observation present.

---

I have changed the message when we mask the browser observations as suggested by @li-boxuan. 

@csmith49 to handle this condenser using ```ObservationMaskingCondenser```, we would need to add a parameter which stores a list of observation classes we want to mask or we need to keep a boolean flag specific for browser outputs and integrate logic of browser condenser into ```ObservationMaskingCondenser``` whenever this flag is true. The default behaviour of ```ObservationMaskingCondenser``` is to mask all observations irrespective of their type.

---

@csmith49 sorry for the ping. Do these changes look good? Any other suggestions/comments?

---

No need to apologize, this fell off my radar and I appreciate the ping.

I don't think we need to extend configuration much in that context: we use the same logic to see if an observation should be masked, and if it should we just apply a hard-coded rule for that kind of observation (defaulting to `<MASKED>` if there is no rule found).

Is the intent here to mask _just_ the browser outputs and leave everything else untouched? I'm trying to think of a scenario where we'd want to have just one kind of observation masked and others ignored.

---

Yes, the intent here is to mask just the browser outputs and leave everything else untouched. This is important because currently we provide screenshots and accessibility trees as input to the model for browser observations. These are really large and consume a lot of token count without any benefits in performance. So we want to mask all such observations from all previous timesteps. Apparently for coding tasks, having all previous observations is helpful it seems.

---

I think I see: you're interested in maintaining just the most recent screenshot and accessibility tree, as opposed to all observations in the attention window. Is that right? If so, I'd suggest updating the condenser doc-string with a bit more detail, as the distinction isn't obvious to me by just looking at the code.